qubes/ext/core_features.py

```diff
@@ -104,7 +104,90 @@ async def qubes_features_request(self, vm, event, untrusted_features):
             untrusted_value = untrusted_features["qubes-agent-version"]
             if _version_re.fullmatch(untrusted_value):
                 vm.features["qubes-agent-version"] = untrusted_value
+
+        # handle boot mode advertisement
+        old_bootmode_info = {}
+        for feature_key, feature_val in vm.features.items():
+            if feature_key.startswith(
+                "boot-mode.kernelopts."
+            ) or feature_key.startswith("boot-mode.name"):
+                old_bootmode_info[feature_key] = feature_val
+        new_bootmode_info = {}
+        untrusted_sanitized_features = {}
+        for (
+            untrusted_feature_key,
+            untrusted_feature_value,
+        ) in untrusted_features.items():
+            if not all(c in string.printable for c in untrusted_feature_value):
+                continue
+            untrusted_sanitized_features[untrusted_feature_key] = (
+                untrusted_feature_value
+            )
+        for (
+            untrusted_feature_key,
+            untrusted_feature_value,
+        ) in untrusted_sanitized_features.items():
+            if untrusted_feature_key.startswith("boot-mode.kernelopts."):
+                bootmode_name = untrusted_feature_key.split(".")[2]
```

Comment:
what if the boot mode id has dots? I don't see them being prohibited... maybe they should be?

Reply:
Hmm, that's a good question. The spec says they should be, but that was originally because I was putting the name in the middle, like `boot-mode.vmreq.kernelopts`. Switching that around to `boot-mode.kernelopts.vmreq` means that a boot mode named `boot-mode.kernelopts....` or `boot-mode.kernelopts.a.b.c` or even just `boot-mode.kernelopts.` should be handled gracefully (though I'd need to test this). I'm not sure whether that qualifies as a bug or a feature, which would you prefer?

Reply:
IMO it will be easier if dots are prohibited in boot mode id.

Reply:
Still need to fix this, will do that on the next iteration.

Reply:
Fixed.